File: app/src/main/java/com/skeensystems/colorpicker/database/ColourDAO.java
package com.skeensystems.colorpicker.database;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;

import java.util.List;

@Dao
public interface ColourDAO {
    @Query("SELECT * FROM SavedColour")
    List<SavedColourEntity> getAll();

    @Query("SELECT * FROM SavedColour WHERE id IN (:colourIds)")
    List<SavedColourEntity> loadAllByIds(int[] colourIds);

    @Query("SELECT * FROM SavedColour WHERE r = :r AND g = :g AND b = :b LIMIT 1")
    SavedColourEntity findByColour(int r, int g, int b);

    @Query("SELECT * FROM SavedColour WHERE favorite LIKE :favorite")
    List<SavedColourEntity> loadFavoriteColours(boolean favorite);

    @Query("UPDATE SavedColour SET favorite = :favorite WHERE id = :id")
    void updateFavoriteColour(long id, boolean favorite);

    @Query("UPDATE SavedColour SET r = :r, g = :g, b = :b WHERE id = :id")
    void updateColour(long id, int r, int g, int b);

    @Insert
    void insertAll(SavedColourEntity... savedColours);

    @Delete
    void delete(SavedColourEntity savedColour);

    @Query("DELETE FROM SavedColour WHERE id = :id")
    void deleteById(int id);
}

File: app/src/main/java/com/skeensystems/colorpicker/database/ColourDAO.kt
package com.skeensystems.colorpicker.database

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.Query

@Dao
interface ColourDAO {
    @Query("SELECT * FROM SavedColour")
    fun getAll(): List<SavedColourEntity>

    @Query("SELECT * FROM SavedColour WHERE r = :r AND g = :g AND b = :b LIMIT 1")
    fun findByColour(
        r: Int,
        g: Int,
        b: Int,
    ): SavedColourEntity

    @Query("SELECT * FROM SavedColour WHERE favorite LIKE :favorite")
    fun loadFavoriteColours(favorite: Boolean): List<SavedColourEntity>

    @Query("UPDATE SavedColour SET favorite = :favorite WHERE id = :id")
    fun updateFavoriteColour(
        id: Long,
        favorite: Boolean,
    )

    @Query("UPDATE SavedColour SET r = :r, g = :g, b = :b WHERE id = :id")
    fun updateColour(
        id: Long,
        r: Int,
        g: Int,
        b: Int,
    )

    @Insert
    fun insertAll(vararg savedColours: SavedColourEntity)

    @Delete
    fun delete(savedColour: SavedColourEntity)

    @Query("DELETE FROM SavedColour WHERE id = :id")
    fun deleteById(id: Int)
}
